File: app/src/main/java/com/denizd/substitutionplan/SubstRepository.java
package com.denizd.substitutionplan;

import android.app.Application;
import android.os.AsyncTask;

import java.util.List;

import androidx.lifecycle.LiveData;

public class SubstRepository {

    private SubstDao substDao;
    private LiveData<List<Subst>> allSubst;

    public SubstRepository(Application application) {
        SubstDatabase database = SubstDatabase.getInstance(application);
        substDao = database.substDao();
        allSubst = substDao.getAllSubst();
    }

    public void insert(Subst subst) {
        new InsertSubstAsync(substDao).execute(subst);
    }

    public void update(Subst subst) {
        new UpdateSubstAsync(substDao).execute(subst);
    }

    public void delete(Subst subst) {
        new DeleteSubstAsync(substDao).execute(subst);
    }

    public void deleteAllSubst() {
        new DeleteAllSubstAsync(substDao).execute();
    }

    public LiveData<List<Subst>> getAllSubst() {
        return allSubst;
    }

    private static class InsertSubstAsync extends AsyncTask<Subst, Void, Void> {

        private SubstDao substDao;
        private InsertSubstAsync(SubstDao substDao) {
            this.substDao = substDao;
        }

        @Override
        protected Void doInBackground(Subst... substs) {
            substDao.insert(substs[0]);
            return null;
        }
    }

    private static class UpdateSubstAsync extends AsyncTask<Subst, Void, Void> {

        private SubstDao substDao;
        private UpdateSubstAsync(SubstDao substDao) {
            this.substDao = substDao;
        }

        @Override
        protected Void doInBackground(Subst... substs) {
            substDao.update(substs[0]);
            return null;
        }
    }

    private static class DeleteSubstAsync extends AsyncTask<Subst, Void, Void> {

        private SubstDao substDao;
        private DeleteSubstAsync(SubstDao substDao) {
            this.substDao = substDao;
        }

        @Override
        protected Void doInBackground(Subst... substs) {
            substDao.delete(substs[0]);
            return null;
        }
    }

    private static class DeleteAllSubstAsync extends AsyncTask<Void, Void, Void> {

        private SubstDao substDao;
        private DeleteAllSubstAsync(SubstDao substDao) {
            this.substDao = substDao;
        }

        @Override
        protected Void doInBackground(Void... voids) {
            substDao.deleteAllSubst();
            return null;
        }
    }
}

File: app/src/main/java/com/denizd/substitutionplan/SubstRepository.kt
package com.denizd.substitutionplan

import android.app.Application
import android.os.AsyncTask

import androidx.lifecycle.LiveData

public class SubstRepository(application: Application) {

    private val substDao: SubstDao?
    val allSubst: LiveData<List<Subst>>?

    init {
        val database = SubstDatabase.getInstance(application)
        substDao = database?.substDao()
        allSubst = substDao?.allSubst
    }

    fun insert(subst: Subst) {
        InsertSubstAsync(substDao).execute(subst)
    }

    fun update(subst: Subst) {
        UpdateSubstAsync(substDao).execute(subst)
    }

    fun delete(subst: Subst) {
        DeleteSubstAsync(substDao).execute(subst)
    }

    fun deleteAllSubst() {
        DeleteAllSubstAsync(substDao).execute()
    }

    class InsertSubstAsync(substDao: SubstDao?) : AsyncTask<Subst, Void, Void>() {
        val mSubstDao = substDao
        override fun doInBackground(vararg substs: Subst): Void? {
            mSubstDao?.insert(substs[0])
            return null
        }
    }

    class UpdateSubstAsync(substDao: SubstDao?) : AsyncTask<Subst, Void, Void>() {
        private val mSubstDao = substDao
        override fun doInBackground(vararg substs: Subst): Void? {
            mSubstDao?.update(substs[0])
            return null
        }
    }

    class DeleteSubstAsync(substDao: SubstDao?) : AsyncTask<Subst, Void, Void>() {
        private val mSubstDao = substDao
        override fun doInBackground(vararg substs: Subst): Void? {
            mSubstDao?.delete(substs[0])
            return null
        }
    }

    class DeleteAllSubstAsync(substDao: SubstDao?) : AsyncTask<Void, Void, Void>() {
        private val mSubstDao = substDao
        override fun doInBackground(vararg voids: Void): Void? {
            mSubstDao?.deleteAllSubst()
            return null
        }
    }
}